File: java/383_Ransom_Note.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        HashMap<Character, Integer> magazineChars = new HashMap<>();
        for (int i = 0; i < magazine.length(); ++i) {
            char m = magazine.charAt(i);
            int currentCount = magazineChars.getOrDefault(m, 0);
            magazineChars.put(m, currentCount + 1);
        }
        for (int i = 0; i < ransomNote.length(); ++i) {
            char r = ransomNote.charAt(i);
            int currentCount = magazineChars.getOrDefault(r, 0);
            if (currentCount == 0) {
                return false;
            }
            magazineChars.put(r, currentCount - 1);
        }
        return true;
    }
}

File: js/383_Ransom_Note.js
/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 */
var canConstruct = function(ransomNote, magazine) {
    let countChars = new Map();;
    for (i = 0; i < magazine.length; ++i) {
        temp = countChars.get(magazine[i]) === undefined
            ? 0 : countChars.get(magazine[i]);
        countChars.set(magazine[i], temp + 1);  
    }
    for (i = 0; i < ransomNote.length; ++i) {
        let temp = countChars.get(ransomNote[i]);
        if (temp <= 0 || temp === undefined) {
            return false;
        }
        countChars.set(ransomNote[i], countChars.get(ransomNote[i]) - 1);
    }
    return true;
};